packages/application-shell/src/components/application-shell/application-shell.js

```diff
@@ -97,241 +96,253 @@ export const RestrictedApplication = props => (
       }
 
       return (
-        // NOTE: we do not want to load the locale data as long as we do not
-        // know the user setting. This is important in order to avoid flashing
-        // of translated content on subsequent re-renders.
-        // Therefore, as long as there is no locale, the children should consider
-        // using the `isLoading` prop to decide what to render.
-        <AsyncLocaleData locale={user && user.language}>
-          {({ isLoading: isLoadingLocaleData, language, messages }) => (
-            <ConfigureIntlProvider
-              // We do not want to pass the language as long as the locale data
-              // is not loaded.
-              {...(isLoadingLocaleData
-                ? {}
-                : {
-                    // We need to pass the value as `undefined` in case the user
-                    // has no `timeZone` defined so the `defaultProps` in the
-                    // `<Context.Provider>` kick in.
-                    timeZone: user && user.timeZone ? user.timeZone : undefined,
-                    language,
-                    messages: mergeMessages(
-                      messages,
-                      i18n[language],
-                      props.applicationMessages[language]
-                    ),
-                  })}
-            >
-              <SetupFlopFlipProvider
-                user={user}
-                defaultFlags={props.defaultFeatureFlags}
+        <ApplicationStateProvider user={user} environment={props.environment}>
+          {/*
+            NOTE: we do not want to load the locale data as long as we do not
+            know the user setting. This is important in order to avoid flashing
+            of translated content on subsequent re-renders.
+            Therefore, as long as there is no locale, the children should consider using the
+            `isLoading` prop to decide what to render.
+          */}
+          <AsyncLocaleData locale={user && user.language}>
+            {({ isLoading: isLoadingLocaleData, language, messages }) => (
+              <ConfigureIntlProvider
+                // We do not want to pass the language as long as the locale data
+                // is not loaded.
+                {...(isLoadingLocaleData
+                  ? {}
+                  : {
+                      language,
+                      messages: mergeMessages(
+                        messages,
+                        i18n[language],
+                        props.applicationMessages[language]
+                      ),
+                    })}
               >
-                <React.Fragment>
-                  {/* NOTE: the requests in flight loader will render a loading
+                <SetupFlopFlipProvider
+                  user={user}
+                  appEnv={props.environment.env}
```

Comment:
We can now simply pass it as a prop instead of connecting from inside the component